src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -143,14 +143,31 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 self.va_end(args[0].immediate())
             }
             "va_copy" => {
-                let va_list = match (tcx.lang_items().va_list(), &result.layout.ty.sty) {
-                    (Some(did), ty::Adt(def, _)) if def.did == did => args[0].immediate(),
-                    (Some(_), _)  => self.load(args[0].immediate(),
-                                               tcx.data_layout.pointer_align.abi),
-                    (None, _) => bug!("`va_list` language item must be defined")
+                let inner_ty = match tcx.lang_items().va_list() {
+                    Some(did) => tcx.type_of(did),
+                    None => bug!("`va_list` language item must be defined")
                 };
+                let impl_ty = match inner_ty.sty {
```

Comment:
You can use `ty_adt_def` here.

Reply:
I considered it myself, but I thought the code winds up about the same (match against `Some` instead of `ty::Adt`). I'll change it in the next commit.